### PR TITLE
GoogleSignIn - Iteration 1 - SignIn/SingOut/SignInWithCached - Android and iOS

### LINKED ISSUE #961: [FEATURE] OAuth (google)
## 🚀 Feature
We will offer (strongly suggest) players to link their game account with a google one using OAuth.

- If no account is logged, players will be able to choose between create an account or do it later.
- If they choose to do it later, the option to link an account should be in the settings menu (from main screen).

### PR DESCRIPTION
**Pull request by BertovDev**
#961 

## Motivation

Implement GoogleSignIn so players can link their google acoounts to the game.

## Summary of changes

- Added GoogleSignIn plugin
- Created `GoogleSignInController` script to controll the SingIn/SingOut/SingInWithCache
- Added funcionality to the Signin button in TittleScreen
- For now we are storing the google information in the PlayersPrefs. Further Iteration it will be stored in our DB.
- Implemented cancelation method when the request takes too long.
- Make it work in Android and iOS
- Created Google OAuth console credentials and app.
- Added https://github.com/googlesamples/unity-jar-resolver. To maange Android and Ios dependencies

Scripts inside GoogleSignInPlugin folder are part of the GoogleSignIn plugin I did not create them.

More information can be found in the following docs I created:

- [Google SignIn options and SingUp/SignIn sequence diagrams](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/BJvny1jO6)
- [Google SignIn OAauth console and unity config](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/HyRViKaOp)
- [GoogleSignIn iOS configuration](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/ByjGNWeq6)
- [Google OAuth set to production](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/BJEV4iTtp)

## How has this been tested?

This can only be tested in mobile, whether it's a simulator (not unity simulator) or a real phone.

Setup:
1. Either ask a @BertovDev or @Nico-Sanchez to add you in their credentials or create your own.
2. Install JDK [here](https://download.oracle.com/java/22/latest/jdk-22_macos-aarch64_bin.dmg)
3. Replace the google ClientID in client/Assets/Scripts/GoogleSignIn/GoogleSignInController.cs:24
4. export the `GOOGLE_CLIENT_ID` variable with the google clientID (2) before running the backend.
5. Replace the url in client/Assets/Scripts/Utils/ServerUtils.cs:226 with ngrok's url pointing to `localhost:4001` while running backend app.

ℹ️ You can create your own testing credentials following this guide -> https://www.notion.so/lambdaclass/Google-Credentials-172a36447ed447dc9d26c202a0b0bf09

1. Press SignIn, a google popup should appear. Choose your account
6. Press play  you should see your google name displayed in the top left corner in Mainscreen
7. Close the App, re-open it. You should be logged with your last account entered.
8. SignOut. Now if you close and re-open no account should be appear as logged
9. SignIn again with any account you want.

Next steps:

- Add .env for credentials
- Handle Errors
- Relationate Google credentials to user in the DB
- How to re-authenticate the user data in the backend
- Change Domains

## Test additions / changes

List tests added/updated.

## Checklist
- [x] I have tested the changes locally.
- [x] I self-reviewed the changes on GitHub, line by line.
- [ ] Tests have been added/updated.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [x] I have tested the changes in another devices.
  - [x] Tested in iOS.
  - [x] Tested in Android.

### CODE REVIEW

**Review by AngieDutra — CHANGES_REQUESTED** [review]
Nice work! Just a few comments:
- Delete SignInSample folder from the repo
- I recommend creating a folder to organize all the files needed for the sign in to work
- I also got an error building the game in my phone:
Screenshot_20240126_150826_client

**Comment by BertovDev** [comment]
> Nice work! Just a few comments:
> 
>     * Delete SignInSample folder from the repo
> 
>     * I recommend creating a folder to organize all the files needed for the sign in to work
> 
>     * I also got an error building the game in my phone:
>       Screenshot_20240126_150826_client

Update:

I did the build in her phone but with my computer and it works. So its probably some build configuration. I will check it now

**Comment by BertovDev** [comment]
Reminder: This is working fine. I would be good to test it and make a build in another computer to finally check if it is a build configuration issue.

**Comment by AminArria** [comment]
Will this work on Unity deskptop?

File: client/Assets/Scenes/TitleScreen.unity [hunk 1]
```diff
@@ -123,6 +123,56 @@ NavMeshSettings:
...
+  m_Enabled: 1
+  m_EditorHideFlags: 0
+  m_Script: {fileID: 11500000, guid: 787a84722d1f84aa6ab1a3668798e7ca, type: 3}
+  m_Name: 
+  m_EditorClassIdentifier: 
+  userName: {fileID: 1516450510}
+  statusText: {fileID: 445723375}
+  signOutButton: {fileID: 1365717059}
+  signInButton: {fileID: 1100061445}
+  webClientIdGoogle: 529212382177-822ukg0eeufi7pivtk1dpatqveqlqord.apps.googleusercontent.com
```
Comment (Nico-Sanchez) [inline client/Assets/Scenes/TitleScreen.unity:174]: Is this necessary? Can we remove it?

**Comment by AngieDutra** [comment]
I thested this PR, but I could check if I actually signed in or not. The text in title was still showing guest. Also, when I came back to change the server instead of sign out it was still saying sign in

**Review by AngieDutra — COMMENTED** [review]
Awesome! But if I return to title screen the log in UI state is nor preserved
Screenshot_20240514_130908_client
Screenshot_20240514_130921_client

**Review by AngieDutra — APPROVED** [review]
Tested in Android builds and is working as expected. Congrats!

**Review by tkz00 — APPROVED** [review]
Looks good!

**Review by AngieDutra — APPROVED** [review]
Awesome!

**Pull request merged**